crates/runc-shim/src/runc.rs

```diff
@@ -349,6 +349,56 @@ impl ProcessLifecycle<InitProcess> for RuncInitLifecycle {
             })
             .collect())
     }
+
+    #[cfg(target_os = "linux")]
+    async fn pause(&self, p: &mut InitProcess) -> Result<()> {
+        let mut state = p.state.lock().await;
+        match *state {
+            Status::UNKNOWN => Err(other!("cannot pause an unknown process")),
+            Status::CREATED => Err(other!("cannot pause a created process")),
+            Status::RUNNING => {
+                *state = Status::PAUSING;
+                if let Err(e) = self.runtime.pause(p.id.as_str()).await {
+                    *state = Status::RUNNING;
+                    return Err(runtime_error(&self.bundle, e, "OCI runtime pause failed").await);
+                }
+                *state = Status::PAUSED;
+                Ok(())
+            }
+            Status::STOPPED => Err(other!("cannot pause a stopped process")),
+            Status::PAUSED => Err(other!("cannot pause a paused process")),
+            Status::PAUSING => Err(other!("cannot pause a pausing process")),
+        }
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    async fn pause(&self, _p: &mut InitProcess) -> Result<()> {
+        Err(Error::Unimplemented("pause".to_string()))
+    }
+
+    #[cfg(target_os = "linux")]
+    async fn resume(&self, p: &mut InitProcess) -> Result<()> {
+        let mut state = p.state.lock().await;
+        match *state {
+            Status::UNKNOWN => Err(other!("cannot resume an unknown process")),
+            Status::CREATED => Err(other!("cannot resume a created process")),
+            Status::RUNNING => Err(other!("cannot resume a running process")),
+            Status::STOPPED => Err(other!("cannot resume a stopped process")),
+            Status::PAUSED => {
+                if let Err(e) = self.runtime.resume(p.id.as_str()).await {
+                    return Err(runtime_error(&self.bundle, e, "OCI runtime pause failed").await);
```

Comment:
`map_err` is more canonical.

Reply:
`runtime_error` is an async function, cannot used by map_err.

Reply:
Right, but `.await` returns `Result`, so you can do `.await.map_err(...)`?

Reply:
Do you mean `self.runtime.resume(p.id.as_str()).await.map_err(runtime_error(...))?`?
This will not pass compilation because the `runtime_error` function is an async function and cannot be used in `map_err`